File: built-in plugins dsp/mb-mic-u87.cpp
/**
 * MB U87 Modeler
 * Category : effect
 * Type     : microphone
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Neumann U87 emulation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIC_U87_H
#define MB_MIC_U87_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMicU87 : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mic-u87";
    static constexpr const char* PLUGIN_NAME    = "MB U87 Modeler";
    static constexpr const char* PLUGIN_TYPE    = "microphone";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float pattern = 0.5f;  // range [0, 1]
    float presence = 0.5f;  // range [0, 1]
    float proximity = 0.3f;  // range [0, 1]
    };

    MbMicU87() = default;
    ~MbMicU87() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.pattern = std::clamp(params.pattern, 0f, 1f);
        params.presence = std::clamp(params.presence, 0f, 1f);
        params.proximity = std::clamp(params.proximity, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB U87 Modeler
        return input;
    }
};

#endif // MB_MIC_U87_H

File: maxbooster7.5/built-in plugins dsp/mb-mic-u87.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMicU87Plugin: PluginDefinition = {
  id: "mb-mic-u87",
  slug: "mb-mic-u87",
  name: "MB U87 Modeler",
  category: "effect",
  type: "microphone",
  version: "1.0.0",
  description: "Neumann U87 emulation",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "pattern",
      name: "Pattern",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "presence",
      name: "Presence",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "proximity",
      name: "Proximity",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { pattern: 0.5, presence: 0.5, proximity: 0.3 },
};

export default MbMicU87Plugin;
